src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1442,7 +1442,10 @@ protected unsafe void WriteEventWithRelatedActivityIdCore(int eventId, Guid* rel
                     if (ex is EventSourceException)
                         throw;
                     else
-                        ThrowEventSourceException(m_eventData[eventId].Name, ex);
+                    {
+                        ref EventMetadata metadata = ref CollectionsMarshal.GetValueRefOrNullRef(m_eventData, eventId);
```

Comment:
Is bypassing the bounds check necessary? If the main improvement to access struct data by reference, can be just `ref m_eventData[eventId]` ?

Replacing safe array access with unsafe is against https://github.com/dotnet/runtime/issues/94941